### PR TITLE
Fixed bug in optimizer rule bitmap

### PR DESCRIPTION
**Pull request by GustavoAngulo**
This PR introduces a fix for #1234. The `bitset.set()` function was being used incorrectly, and not correctly updating the bitmap, leading to false positives. I've also included a test case to check correctness, which without the fix, would highlight the false positive.

### CODE REVIEW

**Comment by malin1993ml** [comment]
@GustavoAngulo @chenboy This PR is not passing the Travis test. Please fix the failed test cases before merging.

**Comment by GustavoAngulo** [comment]
Bugs should be fixed by #1242, that PR should be merged first.

**Comment by malin1993ml** [comment]
Okay I'm reviewing that PR.

File: src/optimizer/memo.cpp [hunk 1]
```diff
@@ -43,8 +43,6 @@ GroupExpression *Memo::InsertExpression(std::shared_ptr<GroupExpression> gexpr,
   auto it = group_expressions_.find(gexpr.get());
 
   if (it != group_expressions_.end()) {
-    PL_ASSERT(target_group == UNDEFINED_GROUP ||
-           target_group == (*it)->GetGroupID());
```
Comment (GustavoAngulo) [inline src/optimizer/memo.cpp]: I debugged for a while and reached the conclusion that this assert was incorrect. From my understanding, it doesn't make sense that if we create a group expression, we search for it in the memo table, and we find an existing equal expression, that we should have the same `GroupID`. Consider two possible join trees (AxB)x(CxD) and Ax(Bx(CxD)). In these two cases, both of the cases will generate a group expression for (CxD) independently.

File: src/optimizer/memo.cpp [hunk 2]
```diff
@@ -43,8 +43,6 @@ GroupExpression *Memo::InsertExpression(std::shared_ptr<GroupExpression> gexpr,
   auto it = group_expressions_.find(gexpr.get());
 
   if (it != group_expressions_.end()) {
-    PELOTON_ASSERT(target_group == UNDEFINED_GROUP ||
-           target_group == (*it)->GetGroupID());
```
Comment (GustavoAngulo) [inline src/optimizer/memo.cpp]: I debugged for a while and reached the conclusion that this assert was incorrect. From my understanding, it doesn't make sense that if we create a group expression, we search for it in the memo table, and we find an existing equal expression, that we should already know it's `GroupID` when its passed in as `target_group`. Consider two possible join trees (AxB)x(CxD) and Ax(Bx(CxD)). In these two cases, both of the cases will generate a group expression for (CxD) independently. The first will generate it from the group Bx(CxD), but the second will generate it from the group (CxD).

Comment (chenboy) [reply src/optimizer/memo.cpp]: Yes after thinking I think you're correct. A case would be from some transformation we get `Ax(BxC)` and insert it into the memo, then from another transformation we get `(AxB)xC` and decide it should go into another group but actually they should be in the same group. Notice that we do not generate all logical expressions at once now so this is possible. Later we'll get this assert failure after applying`Ax(BxC) -> (AxB)xC`. So to solve this we need to merge groups in the memo.

Comment (malin1993ml) [reply src/optimizer/memo.cpp]: I actually think the assertion is fine... I think the fundamental problem is what you have discussed in Slack: the logical and physical join for the same children should be evaluated to be equal. But maybe i’m missing something. Let’s discuss today.

File: src/optimizer/group_expression.cpp [hunk 3]
```diff
@@ -87,15 +87,21 @@ hash_t GroupExpression::Hash() const {
...
+
+  for (size_t i = 0; i < left_groups.size(); ++i) {
+    eq = eq && (left_groups[i] == right_groups[i]);
   }
 
   return eq;
 }
 
 void GroupExpression::SetRuleExplored(Rule *rule) {
-  rule_mask_.set(rule->GetRuleIdx()) = true;
```
Comment (zhaoguoquan94) [inline src/optimizer/group_expression.cpp]: In #1285 , you mentioned that  
>We created and added a handful of "plan_selection_tests" that test if the optimizer does the appropriate join ordering (it currently does not).

Is this line the reason that the optimizer is not doing it? Or there are other rules that cause this problem?  
For simple join, say 2 tables, is it possible that the "logical to physical" rule is applied first and the rest of the possible plan are pruned off, like early stopping?

Comment (GustavoAngulo) [reply src/optimizer/group_expression.cpp]: I am not sure about possible early pruning, however, without this change, rules such as Join Commutativity were never being applied

**Pull request merged**